enhancements/cvo/cluster-profiles.md

```diff
@@ -0,0 +1,103 @@
+---
+title: Cluster Profiles
+authors:
+  - "@csrwng"
+reviewers:
+  - "@abhinavdahiya"
+  - "@crawford"
+  - "@sdodson"
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+approvers:
+  - "@derekwaynecarr"
+  - "@smarterclayton"
+creation-date: 2020-02-04
+last-updated: 2019-02-04
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# cluster-profiles
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+Cluster profiles are a way to support different deployment models for OpenShift clusters. 
+A profile is an identifier that the Cluster Version Operator uses to determine
+which manifests to apply. Operators can be excluded completely or can have different
+manifests for each supported profile.
+
+## Motivation
+
+In order to support different deployment models in which not all operators rendered by
+the CVO by default are needed, a mechanism is needed by which we tell the CVO which 
+manifests to include or exclude. One such deployment is the hosted control plane case
+in which the control plane components and corresponding operators should not be rendered
+by the CVO.
+
+
+### Goals
+
+- Equip the CVO to include/exclude manifests based on a matching identifier (profile)
+
+### Non-Goals
+
+- Define which manifests should be excluded/included for each use case
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+As a user, I can create a cluster in which manifests for control plane operators are
+not applied by the CVO.
+
+#### Story 2
+As a user, I can create a cluster in which node selectors for certain operators target
+worker nodes instead of master nodes.
+
+### Design
+
+A cluster profile is specified to the CVO as an identifier in an environment
+variable. For a given cluster, only one CVO profile may be in effect.
+
+```
+CLUSTER_PROFILE=[identifier]
+```
+This environment variable would have to be specified in the CVO deployment. When
+no `CLUSTER_PROFILE=[identifier]` variable is specified, the default cluster profile
+is in effect.
+
+The following annotations may be used to include/exclude manifests for a given profile:
+
+```
+exclude.release.openshift.io/[identifier]=true
+```
+This would exclude the manifest from rendering when `CLUSTER_PROFILE=[identifier]`
+has been specified. For the default cluster profile, manifests that contain only
+exclude annotations are always included.
+
+```
+include.release.openshift.io/[identifier]=false
+```
+This would make the CVO render this manifest only when `CLUSTER_PROFILE=[identifier]`
+has been specified. For the default cluster profile (no `CLUSTER_PROFILE` variable), 
+manifests that have any include annotation will always be excluded.
```

Comment:
It would be good to cover how we expect cluster profiles to be exposed to users other than this environment variable - i.e. I presume in most cases we're not expecting customers to modify the CVO deployment to add the env var

e.g. would this be an install-config.yaml thing, templated into bootkube.sh, passed to `cvo render`, and templated into the CVO deployment?

Reply:
@markmc the initial use case scenario was for ROKS offering on IBM Cloud.  I would want to find a model that worked there first before expanding scope much more.  If we want to think about two use cases, CRC (excluding HA deployment topology versus single node concerns) maybe a good counter example.

Reply:
That's fine ... it's just entirely unclear to me how we expect this env variable to be set in the hypershift, CRC, or any other case. Maybe I'm missing something obvious.

Reply:
I left that part undefined (and now adding text that explicitly says so) because potential use cases such as hypershift and CRC take completely different approaches to getting a useable cluster up and running. Both bypass the regular installer in some form, but describing how that should be done would confuse this proposal.

Reply:
For CRC, we are currently leaning towards a configmap containing the cluster information which is set at install time https://github.com/openshift/cluster-version-operator/pull/404